src/cas.c

```diff
@@ -0,0 +1,34 @@
+#include "writer.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+
+bool __tivo_sync_bool_compare_and_swap_4(volatile int32_t *data,
+                                         volatile int32_t expected,
+                                         volatile int32_t desired) {
+  bool use_replay = getenv(RR_MODE_ENVVAR) == EXEC_MODE_REPLAY;
```

Comment:
`getenv` returns a string (`char*`) so you cannot directly compare that to `EXEC_MODE_REPLAY` which expands to an integer.

```suggestion
  bool use_replay = atoi(getenv(RR_MODE_ENVVAR)) == EXEC_MODE_REPLAY;
```

The easiest fix would be to use `atoi` to convert the `char*` to an int but it causes an error if the environment variable is not set.

```suggestion
  char *mode = getenv(RR_MODE_ENVVAR);
  bool use_replay = false; // record by default
  if (mode != NULL) {
    // if `RR_MODE_ENVVAR` is not valid, `atoi` returns 0 so use_replay will be false
    use_replay = atoi(getenv(RR_MODE_ENVVAR)) == EXEC_MODE_REPLAY;
  }
```

Another thing to note is `atoi` returns 0 if the input is not convertible to a number.

Reply:
good call out. I decided it's probably worth wrapping this functionality in a library function.